src/backend/halo2.rs

```diff
@@ -305,3 +310,110 @@ pub fn to_halo2_advice<F: Field>(
         _ => panic!("jarll wrong phase"),
     }
 }
+
+#[derive(Clone, Default)]
+pub struct ChiquitoHalo2Circuit<F: Field + From<u64>> {
+    compiled: ChiquitoHalo2<F>,
+    witness: Option<Assignments<F>>,
+}
+
+impl<F: Field + From<u64>> ChiquitoHalo2Circuit<F> {
+    pub fn new(compiled: ChiquitoHalo2<F>, witness: Option<Assignments<F>>) -> Self {
+        Self { compiled, witness }
+    }
+}
+
+impl<F: Field + From<u64> + Hash> h2Circuit<F> for ChiquitoHalo2Circuit<F> {
+    type Config = ChiquitoHalo2<F>;
+
+    type FloorPlanner = SimpleFloorPlanner;
+
+    type Params = ChiquitoHalo2<F>;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn params(&self) -> Self::Params {
+        self.compiled.clone()
+    }
+
+    fn configure_with_params(
+        meta: &mut ConstraintSystem<F>,
+        mut compiled: Self::Params,
+    ) -> Self::Config {
+        compiled.configure(meta);
+
+        compiled
+    }
+
+    fn synthesize(
+        &self,
+        compiled: Self::Config,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        compiled.synthesize(&mut layouter, self.witness.as_ref());
+
+        Ok(())
+    }
+
+    fn configure(_: &mut ConstraintSystem<F>) -> Self::Config {
+        unreachable!()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct ChiquitoHalo2SuperCircuit<F: Field + From<u64>> {
+    sub_circuits: Vec<ChiquitoHalo2<F>>,
+    witness: SuperAssignments<F>,
+}
+
+impl<F: Field + From<u64>> ChiquitoHalo2SuperCircuit<F> {
+    pub fn new(sub_circuits: Vec<ChiquitoHalo2<F>>, witness: SuperAssignments<F>) -> Self {
+        Self {
+            sub_circuits,
+            witness,
+        }
+    }
+}
+
+impl<F: Field + From<u64> + Hash> h2Circuit<F> for ChiquitoHalo2SuperCircuit<F> {
+    type Config = Vec<ChiquitoHalo2<F>>;
+
+    type FloorPlanner = SimpleFloorPlanner;
+
+    type Params = Vec<ChiquitoHalo2<F>>;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn params(&self) -> Self::Params {
+        self.sub_circuits.clone()
+    }
+
+    fn configure_with_params(
+        meta: &mut ConstraintSystem<F>,
+        mut sub_circuits: Self::Params,
+    ) -> Self::Config {
+        sub_circuits.iter_mut().for_each(|c| c.configure(meta));
+
+        sub_circuits
+    }
+
+    fn synthesize(
+        &self,
+        sub_circuits: Self::Config,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        for sub_circuit in sub_circuits {
+            sub_circuit.synthesize(&mut layouter, self.witness.get(&sub_circuit.ir_id))
+        }
```

Comment:
If we get witness using ir_id, does it mean the SuperAssignments HashMap can only have one assignment for each ir_id? What about circumstances where we want to plugin multiple witnesses for the same subcircuit? Do we plug in a different SuperAssignments entirely?